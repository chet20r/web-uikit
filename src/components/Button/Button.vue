<template>
  <component
    :is="tag"
    class="btn"
    :class="[theme, { outline, rounded, disabled, fullwidth }, size, ...styles]"
    :type="type"
    v-bind="$attrs"
    @click="$emit('click', $event)"
  >
    <slot name="left"></slot>
    <slot></slot>
    <slot name="right"></slot>
  </component>
</template>

<script>
import propMixins from '@/components/mixins/props';
export default {
  name: 'Button',
  mixins: [propMixins],
  props: {
    tag: {
      required: false,
      type: String,
      default: 'button',
      validator: tag => tag.match(/(button|a)/)
    },
    type: {
      required: false,
      type: String,
      default: 'button',
      validator: type => type.match(/(button|submit|reset)/)
    },
    disabled: {
      required: false,
      type: Boolean,
      default: false
    },
    theme: {
      required: false,
      type: String,
      default: 'primary',
      validator: theme => theme.match(/(primary|secondary)/)
    },
    rounded: {
      required: false,
      type: Boolean,
      default: false
    },
    outline: {
      required: false,
      type: Boolean,
      default: false
    },
    fullwidth: {
      required: false,
      type: Boolean,
      default: false
    },
    size: {
      required: false,
      type: String,
      default: 'md',
      validator: size => size.match(/(sm|md|lg|xl)/)
    }
  }
};
</script>

<style lang="postcss">
.btn {
  @apply px-2 py-2
    border-2 border-solid 
    uppercase font-anko-bold
    rounded shadow;

  &.sm {
    @apply text-sm px-4 py-1;
  }
  &.md {
    @apply text-base;
  }
  &.lg {
    @apply text-lg;
  }
  &.xl {
    @apply text-xl;
  }

  &.fullwidth {
    @apply w-full;
  }

  &:focus {
    @apply outline-none shadow-outline;
  }

  &.rounded {
    @apply px-4 rounded-full;
  }

  &.primary {
    @apply border-blue bg-blue text-white;
    &:hover {
      @apply bg-white text-blue;
    }
    &.outline {
      @apply text-blue bg-white;
      &:hover {
        @apply bg-blue text-white;
      }
    }
  }

  &.secondary {
    @apply border-orange bg-orange text-white;
    &:hover {
      @apply bg-white text-orange;
    }
    &.outline {
      @apply text-orange bg-white;
      &:hover {
        @apply bg-orange text-white;
      }
    }
  }
}
</style>
