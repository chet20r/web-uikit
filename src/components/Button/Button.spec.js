import { shallowMount, mount } from '@vue/test-utils';
import Button from './Button';

describe('Button', () => {
  let wrapper;
  beforeAll(() => {});

  it('is a Vue instance', () => {
    wrapper = shallowMount(Button);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it('should render button element by default', () => {
    wrapper = shallowMount(Button, {
      slots: {
        default: 'Button Text'
      }
    });
    expect(wrapper.contains('button')).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render anchor element when tag props is set to `a`', () => {
    wrapper = shallowMount(Button, {
      propsData: {
        tag: 'a'
      }
    });
    expect(wrapper.contains('a')).toBe(true);
  });

  it('should render the button text in default slot', () => {
    wrapper = shallowMount(Button, {
      slots: {
        default: 'Default Button'
      }
    });
    expect(wrapper.text()).toBe('Default Button');
  });

  it('should have default `btn` class', () => {
    wrapper = shallowMount(Button);
    expect(wrapper.classes()).toContain('btn');
  });

  it('should have `primary` class when passed as theme prop', () => {
    wrapper = shallowMount(Button, {
      propsData: {
        theme: 'primary'
      }
    });
    expect(wrapper.classes()).toContain('primary');
  });

  it('should have classes added for boolean props', () => {
    wrapper = shallowMount(Button, {
      propsData: {
        rounded: true,
        inverted: true,
        fullwidth: true
      }
    });
    expect(wrapper.classes()).toContain('rounded', 'inverted', 'fullwidth');
  });

  it('should have class md as default size', () => {
    wrapper = shallowMount(Button, {});
    expect(wrapper.classes()).toContain('md');
  });

  it('should have override style classes passed in style prop', () => {
    wrapper = shallowMount(Button, {
      propsData: {
        styles: ['a', 'b']
      }
    });
    expect(wrapper.classes()).toContain('a', 'b');
  });

  it('should have class `lg` when setting size prop', () => {
    wrapper = shallowMount(Button, {
      propsData: {
        size: 'lg'
      }
    });
    expect(wrapper.classes()).toContain('lg');
  });

  it('should have non prop attributes passed through', () => {
    wrapper = shallowMount(Button, {
      attrs: {
        custom: 'attr'
      }
    });
    expect(wrapper.attributes('custom')).toBe('attr');
  });
});
