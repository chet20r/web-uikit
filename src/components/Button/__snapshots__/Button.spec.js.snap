// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button element by default 1`] = `
<button
  class="btn primary md"
  type="button"
>
   Button Text 
</button>
`;
